### ISSUE TITLE
Add more annotations to twisted.internet

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#9926]
|[<img alt="mthuurne's avatar" src="https://avatars.githubusercontent.com/u/246676?s=50" width="50" height="50">](https://github.com/mthuurne)| @mthuurne reported|
|-|-|
|Trac ID|trac#9926|
|Type|enhancement|
|Created|2020-07-30 10:25:35Z|

Since mypy only checks annotated function/method bodies, we have to add annotations to increase the coverage of mypy's checks, which in turn helps validate the annotations that are already there.

I've been doing some more automatic annotation adding and these are the issues I'm running into:

## Missing arguments to reactor methods

Unit tests call reactor methods, in particular `listenX()` and `connectX()` methods, without a full argument list. Because the reactor interfaces do not contain argument defaults, mypy considers this an error. See [#9911](https://github.com/twisted/twisted/issues/9911).

## Test result lists

A common pattern found in Twisted's unit test is this:

```
    wasCalled = []
    def callback():
        wasCalled.append(None)
    self.assertTrue(wasCalled)
```

Since mypy wants to know the element type for all lists, a construct like this would require an additional `# type: List[None]` annotation.

In modern Python, we have the `nonlocal` keyword, so this can be rewritten as:

```
    wasCalled = False
    def callback() -> None:
        nonlocal wasCalled
        wasCalled = True
    self.assertTrue(wasCalled)
```

So in this case, the solution is clear, but a lot of code needs to change.

## Understanding `ReactorBuilder`

`ReactorBuilder.buildReactor()` will return a reactor that implements the interfaces specified in the `requiredInterfaces` class variable. However, mypy doesn't know this and will flag any use of methods outside of IReactorCore as an error.

This can be worked around using typecasts or statements like `assert isinstance(reactor, IReactorTime)`, but since there are a lot of test cases, that requires a lot of extra code only for the benefit of mypy.

Therefore I've written a mypy plugin to teach it about `ReactorBuilder` and automatically adjust the return type of `buildReactor()` to match `requiredInterfaces`. It's mostly working, but the typing of `requiredInterfaces` is a problem: if it's a list, I can't seem to get its values from mypy (since they're not types), while if it's a tuple, a class inheriting from multiple mixins will error out on having conflicting types for `requiredInterfaces` among its base classes. While it is possible to annotate in a way to avoid or suppress that error, I haven't found a way of doing so that still provides the plugin with the information it needs.

I'm now considering changing `requiredInterfaces` from a class variable to a decoration.

## Type-checking mixins

A mixin can access variables or methods that don't exist in the mixin itself, but are expected to be present once the mixin has been mixed into a concrete class. Since mypy doesn't know this, it will report errors for every non-existing class member.

One way around this is to declare the expected members on the mixin:

```
class TestMixin:
    if TYPE_CHECKING:
        def doTheThing(self) -> None: ...
```

Alternatively, this methods could be declared as an abstract methods without a `TYPE_CHECKING` guard, but then it is important that the mixin comes after the implementation class in the method resolution order.

This approach works fine if there is a handful of expected members specific to a particular mixin. However, for unit tests, the `unittest.TestCase` class contains lots of methods like `assertEqual()` and repeating all of those in mixins would become a mess.

It is possible to use protocols and annotation of `self` to annotate mixins, as explained in [mypy's documentation](https://mypy.readthedocs.io/en/latest/more_types.html#mixin-classes). That would require writing a protocol for every concrete class we want to mix with. So for `unittest.TestCase` that would still be a rather large amount of code, but it would have to be done only once for all of Twisted instead of once per mixin. It also means we'd have to annotate `self` for every test method.

Since `typing.Protocol` was only added to the standard library in Python 3.8, we'd have to use the `typing-extensions` package to get support for it in older Python versions.

Another workaround is to have the mixin not be a mixin during type checking:

```
if TYPE_CHECKING:
    MixinBase = unittest.TestCase
else:
    MixinBase = object

class TestMixin(MixinBase):
    def test_something(self) -> None:
        self.assertEqual(1 + 1, 2)
```

This works reasonably well, but it does require some tweaking of the order of base classes to get a consistent method resolution order both during type checking and at runtime.

I'm also considering a plugin approach here: by decorating a mixin class, we can let the plugin know about the extra variables and methods we expect to be present after mixing.

```
@mixin(unittest.TestCase)
class TextMixin:
    def test_something(self) -> None:
        self.assertEqual(1 + 1, 2)
```

The implementation of the `mixin` decorator doesn't have to do anything at runtime: it's only there to pass metadata to the plugin.

## Impact for projects type-checking against Twisted

There is also the question of how the solution that we use for Twisted internally affects external projects that want to type-check their code against an installed version of Twisted.

Since `ReactorBuilder` is only used by unit tests, it is not part of Twisted's API, so whatever solution we pick there won't affect external projects.

The majority of mixins are used in unit tests, but there are a few in public API modules as well. So if we decide to use `typing-extensions` or a custom plugin, we'd either have to restrict the use of that solution to unit tests only, or require that external projects also use `typing-extensions` or the plugin.

Note that almost any project type-checking versus Twisted already needs to use the `mypy-zope` plugin, so requiring a plugin may not be too much of an ask. However, in that case the plugin should be part of the distribution (wheel) instead of living among the internal `admin` scripts.

There is also the open issue of whether `Deferred` can be supported by standard syntax only or will require a mypy plugin, see [#9816](https://github.com/twisted/twisted/issues/9816).

<details><summary>Searchable metadata</summary>

```
trac-id__9926 9926
type__enhancement enhancement
reporter__mthuurne mthuurne
priority__normal normal
milestone__publish_types publish types
branch__ 
branch_author__ 
status__new new
resolution__None None
component__core core
keywords__None None
time__1596104735810877 1596104735810877
changetime__1596104735810877 1596104735810877
version__None None
owner__None None

```
</details>
